### DISCUSSION TITLE
Nested Json

### DISCUSSION THREAD

**Discussion by alexandra-buzatoiu** [nlohmann/json#3761] [Q&A]
Is there a way I can do some nested Jsons?
Basically, I have a class and I want to implement fromJson and toJson Methods, in order to put my entire object inside a json, like simulating the c# JsonConvert.SerializeObject(myobject)
My class looks like this:
class Blockchain {

	int difficulty = 2;
	std::vector<Block> Chain;
	int Reward;

and it consists of a vector of other defined objects and some other integers. The Block looks like this:

class Block {

	int Index;
	time_t TimeStamp;
	std::string PreviousHash;
	std::string Hash;
	std::string Data;
	int nounce = 0;

I have managed to implement the functions fromJson and toJson for Block like this:

void Block::to_json(json& j)
{
	j = json{ {"Index", this->Index}, {"TimeStamp", this->TimeStamp}, {"PreviousHash", this->PreviousHash}, {"Hash", this->Hash}, {"Data", this->Data}, {"nounce", this->nounce} };
}

void Block::from_json(const json& j)
{
	j.at("Index").get_to(this->Index);
	j.at("TimeStamp").get_to(this->TimeStamp);
	j.at("PreviousHash").get_to(this->PreviousHash);
	j.at("Hash").get_to(this->Hash);
	j.at("Data").get_to(this->Data);
	j.at("nounce").get_to(this->nounce);
}

When I wanted to do the same thing for the Blockchain class, it worked if I ignored the integers and only work with the vector, and I had it implemented like this:

void Blockchain::to_json(json& j)
{

	for (int i = 0; i < this->Chain.size(); i++)
	{
		json ja;
		(this->Chain.begin()+i)->to_json(ja);
		j.push_back(ja);
		
	}
	
}

void Blockchain::from_json(const json& j)
{

	for (auto& el : j.items())
	{
		json s = el.value();
		Block b(s);
		this->AddBlock(b);
		
	}
}

My question is, what should I do in order to be able to write toJson and fromJson functions also with another members apart from the same elements of a vector. 

I tried something like this but didn't work.

void Blockchain::to_json(json& j)
{

        json aux;
	for (int i = 0; i < this->Chain.size(); i++)
	{
		json ja;
		(this->Chain.begin()+i)->to_json(ja);
		jaux.push_back(ja);
		
	}
       j.push_back(jaux);
	
}

Is there a simpler method to do this? Thank you in advance!

**Comment by falbrechtskirchinger** [comment]
The proper way of serializing objects is described here: https://json.nlohmann.me/features/arbitrary_types/

This also enables automatic handling of vectors of serializable types, etc.

**Reply by alexandra-buzatoiu** [reply]
I understood this, but still don't get if it is possible to create a json of an object which contains some other things than a vector (which I have seen I can serialize and turn into a json). So my approach would be to create the jsons for each element in the vector, and pushback into the big json. But how can I add other items to the big json? Non related to the vector. It might be my fault, but I haven't seen this detail in the documentation.

**Reply by falbrechtskirchinger** [reply]
The same way you add anything to a JSON object?
https://json.nlohmann.me/api/basic_json/operator[]/

(You haven't provided a definition for `Blockchain` btw., so I don't know what you're trying to add specifically.)

For the vector you should be able to do:
```cpp
json ja{this->Chain};
```

Assuming proper conversion functions for `Block` exist (which you haven't shown).
